engine/src/main/scala/cromwell/engine/workflow/lifecycle/WorkflowFinalizationActor.scala

```diff
@@ -90,29 +115,46 @@ case class WorkflowFinalizationActor(workflowTag: String, backendsToFinalize: Ma
     }
   }
 
-  private def checkForDone(stateData: WorkflowFinalizationActorData) = {
-    // TODO: Check these are actually the same?
-    val allActorResponses = stateData.successes ++ stateData.failures.keys
-    allActorResponses.size == backendsToFinalize.size
-  }
+  private def checkForDone(stateData: WorkflowFinalizationActorData) = stateData.finalizationActors.isEmpty
 
   whenUnhandled {
     case unhandledMessage =>
-      log.warning(s"Backends finalizer for $workflowTag received an unhandled message: $unhandledMessage")
+      log.warning(s"$tag received an unhandled message: $unhandledMessage")
       stay
   }
 
   onTransition {
     case _ -> toState if toState.terminal =>
-      log.debug(s"$tag is done. Shutting down.")
+      log.debug(s"$tag state is now terminal. Shutting down.")
       context.stop(self)
     case fromState -> toState =>
-      log.debug(s"$tag is transitioning from $fromState to $toState.")
+      log.debug(s"$tag transitioning from $fromState to $toState.")
   }
 
   /**
-    * Based on the backendAssignments, creates a BackendWorkflowFinalizationActor per backend and maps the actor to the
-    * name of the backend which it is finalizing.
+    * Based on backendAssignments, creates a BackendWorkflowInitializationActor per backend and maps the actor to the
+    * name of the backend which it is initializing.
     */
-  def backendWorkflowFinalizationActors:  Map[ActorRef, String] = ??? //TODO: Implement this!
+  def backendWorkflowFinalizationActors(backendAssignments: Map[Call, String]):  Map[ActorRef, String] = {
+    val callAssignmentMap = callAssignments(backendAssignments)
+    val backendsToFinalize = backendAssignments.values.toSet
+    backendsToFinalize map { backend => (finalizationActor(backend, callAssignmentMap(backend)), backend) } toMap
+  }
+
+  /**
+    * Turns a mapping from Call => backendName into a mapping from backendName => List[Call]
+    */
+  private def callAssignments(backendAssignments: Map[Call, String]): Map[String, Seq[Call]] = {
+    // Group the assignments by backendName:
+    val asTuples = backendAssignments.toSeq
+    val groupedByBackendName: Map[String, Seq[(Call, String)]] = asTuples groupBy { case (call, backendName) => backendName }
```

Comment:
I'd be very interested in a tidier version here. It's like this because there's no groupBy on Map (since Map doesn't extend Seq).
... @mcovarr is usually very good at suggesting tidier versions of my messy code...?